### DISCUSSION TITLE
Using __local memory inside devicelib

### DISCUSSION THREAD

**Discussion by pvelesko** [CHIP-SPV/chipStar#902] [General]
Any suggestions on how to best allocate and pass __local memory to functions inside devicelib kernels? 

Normally, this memory would be allocated inside kernel.

@linehill @pjaaskel

**Comment by linehill** [comment]
Do you mean you want to do something like this in non-kernel function?

```c
void foo(...) {
  static local float lcl_mem[123]; // Note: invalid OpenCL C.
  ...
}
```

**Comment by pvelesko** [comment]
Yes, I would like to use local memory inside device function but the issue is that I can't allocate it inside device function. Was wondering what would be the best way to approach this

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pvelesko** [comment]
From @pjaaskel :

| Can't you just move that local array inside the synchtreads_count() implementation?

No, because you can only allocate memory in kernel regions - so this memory needs to be allocated somewhere else and passed in as argument

**Reply by pjaaskel** [reply]
Ah, I didn't remember that restriction. In PoCL I created an under-the-hood mechanism for allocating local space to implement subgroups for the CPU. If we cannot get around this need, we might need a similar mechanism for chipStar.

**Reply by pvelesko** [reply]
> under-the-hood mechanism for allocating local space 

How does it work? 

> If we cannot get around this need

Currently I need this for implement `__syncthreads_count()`, for example. I'm sure having such feature would be useful in the future.

**Reply by pjaaskel** [reply]
In practice, a magic function call that asks for dynamic (local size dependent) dize local memory and is expanded by the compiler to allocate it. Not sure can it be made work at SPIR-V level in general as I think automatic locals must be compile-time sized.

**Reply by pvelesko** [reply]
@pjaaskel 

> SPIR-V level in general as I think automatic locals must be compile-time sized.

I've found that automatic __local arrays get translated to spirv as just pointers to workgroup memory without any indication of size.
```
__local int var[10]
```

I think I found a solution that works for implementing a single device function that requires `__local` memory (i.e. it's not magic) 

1. Declare `__local` memory as `extern` inside the device function that needs it
2. Compile `device_function.cl` to SPIR-V
3. Edit human readable SPIR-V and delete this line `OpDecorate %localMem LinkageAttributes "localMem" Import`
4. For some reason in step 2 `__local` gets compiled to `CrossWorkgroup` (`__global`) so replace relevant instances of `CrossWorkgroup` with `Workgroup` (`__local`)
5. Assemble back into machine SPIR-V.
6. Use the rtdevlib mechanism to insert this into the runtime. 

This compiles and seems to be giving me a reasonable answer so far (early tests). 

Given this workaround, it seems that allocating `__local` memory in device functions should be possible?

**Reply by pjaaskel** [reply]
I wouldn't draw that conclusion just yet. These type of tricks often work on luck on some systems and fail miserably on others. If there is no indication of size, who will allocate that local memory and ensure you won't overwrite someone else's local memory areas (likely another work-group's) etc.?

**Reply by pvelesko** [reply]
> If there is no indication of size, who will allocate that local memory and ensure you won't overwrite someone else's local memory areas (likely another work-group's) etc.?

I don't understand why there is no indication of size being generated by SPIR-V.

**Reply by linehill** [reply]
| I don't understand why there is no indication of size being generated by SPIR-V.

There should be like in this [example](https://godbolt.org/z/x89dMqEn9) where the size of the `var` is encoded in `%_arr_uint_ulong_10 = OpTypeArray ...`.

**Comment by pvelesko** [comment]
Yeah pekka messaged me yesterday and i got it cleared up.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by pjaaskel** [reply]
https://github.com/KhronosGroup/OpenCL-Docs/issues/1221
